monkey/monkey_island/cc/services/initialize.py

```diff
@@ -0,0 +1,7 @@
+from monkey_island.cc.services.post_breach_files import PostBreachFilesService
+from monkey_island.cc.services.run_local_monkey import RunLocalMonkeyService
+
+
+def initialize_services(data_dir):
```

Comment:
If we need to add more configurable values we'll just keep adding parameters to this method like `def initialize_services(data_dir, island_port, bootloader_port):`? . `PostBreachFilesService` and `RunLocalMonkeyService` doesn't need any special initialization, they just need to access configuration parameters.

Reply:
> If we need to add more configurable values we'll just keep adding parameters to this method like def initialize_services(data_dir, island_port, bootloader_port):?

Yes. If the configuration list gets too long, we can create a config object and pass in the config object, rather than individual parameters. The initialization function can then pass only those things that each service needs to that service.

> PostBreachFilesService and RunLocalMonkeyService doesn't need any special initialization, they just need to access configuration parameters.

Right, but they *only* need access to one specific configuration parameter. By passing them exactly what the need, they become decoupled from global state, easier to test, and their dependencies are clearer.

At the moment, we're hamstrung by the architecture, which is a bunch of classes that pretend to be static and contain no state, but actually do contain state in the form of an environment singleton. These should be instance objects that are passed to the resources via dependency injection or accessed via a service locator.

Reply:
> Yes. If the configuration list gets too long, we can create a config object and pass in the config object, rather than individual parameters. The initialization function can then pass only those things that each service needs to that service.

So what I'm suggesting is decoupling environment config files from the singleton. And that singleton would be the config object.

> Right, but they only need access to one specific configuration parameter. By passing them exactly what the need, they become decoupled from global state, easier to test, and their dependencies are clearer.

So why not do:
```
class PublicVarReader():
    PUBLIC_VAR = env_singleton.public_var
```
This also passes exactly one configuration parameter and when testing you'd mock `PUBLIC_VAR` just the same. If `public_var` is indeed a global config parameter, like island port, it should stay in environment singleton no matter what. What could change is that one day we decide to pull the island port from cmd instead of a file, but that's the responsibility of a singleton builder. Imagine we added the ability to customize the port of an island by specifying it in config file. If you want to create a class that needs to know the island port, you'd first need to create a "class builder" somewhere else in the codebase to create a class itself. That's less intuitive than reading a global value from a singleton and adds unnecessary complexity IMO. 
Nevertheless, if you thought about it and made this design choice with care, I trust that you have a good long term vision for the design and don't require a change.

Reply:
> I trust that you have a good long term vision for the design and don't require a change.
I have a long term plan :)

> So what I'm suggesting is decoupling environment config files from the singleton.

Ultimately, environment config will very likely go away. The deployment type shouldn't really be configurable, at least not in the sense that the log level is configurable. The user password and hash shouldn't be written into a config file, but some other file specifically for tracking that that is read-only. We should, at some point, replace our authentication system with something 3rd party (it would be more secure) that would support multiple users (our users have asked for multi-user support).

> This also passes exactly one configuration parameter and when testing you'd mock PUBLIC_VAR just the same.

Making this global means your class is coupled to the specific details of how configuration is managed. It also means you *need* to mock it; in other words, your tests are also coupled to the specific details of your configuration. Passing a class the values it needs decreases coupling, simplifies testing, and makes the interface explicit. Things that are decoupled are easier to change (see The Pragmatic Programmer, The Essence of Good Design and Decoupling).

Anyway, we can discuss this more on a call (better communication faster).